File: 1514/1514.py
class Solution:
    def maxProbability(self, n: int, edges: List[List[int]], probs: List[float], start: int, end: int) -> float:
        adj = [[] for _ in range(n)]

        for i, (u, v) in enumerate(edges):
            adj[u].append([v, probs[i]])
            adj[v].append([u, probs[i]])

        dp = [0 for _ in range(n)]
        pq = [[-1, start]]

        while pq:
            p, u = heappop(pq)
            if u == end: break
            for v, q in adj[u]:
                if (p2 := p * q) < dp[v]:
                    dp[v] = p2
                    heappush(pq, [p2, v])

        return -dp[end]

File: 1514/1514.cpp
# define VD vector<double>

class Solution {
public:
    double maxProbability(int n, vector<vector<int>>& edges, vector<double>& probs, int start, int end) {
        VD dp(n, 0.0); dp[start] = 1.0;

        while (n--){
            bool valid = false;
            for (int i = 0; i < edges.size(); i++){
                int u = edges[i][0], v = edges[i][1];
                double p = probs[i];
                if (dp[u]*p > dp[v]){
                    dp[v] = dp[u]*p;
                    valid = true;
                }
                if (dp[v]*p > dp[u]){
                    dp[u] = dp[v]*p;
                    valid = true;
                }
            }
            if (!valid) break;
        }

        return dp[end];
    }
};
